include/boost/type_index/runtime_cast.hpp

```diff
@@ -0,0 +1,141 @@
+//
+// Copyright (c) CHris Glover, 2016.
+//
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#ifndef BOOST_TYPE_INDEX_RUNTIME_CAST_HPP
+#define BOOST_TYPE_INDEX_RUNTIME_CAST_HPP
+
+/// \file runtime_cast.hpp
+/// \brief Contains boost::typeindex::ctti_type_index class that is constexpr if C++14 constexpr is supported by compiler.
+///
+/// boost::typeindex::runtime_cast class can be used as a replacement
+/// for dynamic_cast on platforms where dynamic_cast is not available
+/// or undesirable at a global level.
+
+#include <boost/type_index.hpp>
+#include <boost/throw_exception.hpp>
+#include <type_traits>
+#include <typeinfo>
```

Comment:
It's for std::bad_cast which I need for the reference versions (since they can't return nullptr). An alternative is to use a different exception type or move the reference versions to a different header.